Be more lenient about refresh tokens.

rebeckag opened the pull request:
Follow the same implementation details regarding refresh tokens as in the Android SDK, resulting from [this discussion](https://github.com/openid/AppAuth-Android/pull/42).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openid/appauth-ios/23)

<!-- Reviewable:end -->

iainmcgin commented:
+@williamdenniss and +@stevenewright for review.

<details><summary><i>Previously, rebeckag (Rebecka Gulliksson) wrote…</i></summary>
> #### Be more lenient about refresh tokens\.
> 
> Follow the same implementation details regarding refresh tokens as in the Android SDK, resulting from [this discussion](https://github.com/openid/AppAuth-Android/pull/42).
> 
> 
> 
</details>


---

Review status: 0 of 1 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23#-:-KIUOMi7h7ZkaLnqCn1S:702100335)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
Sorry for the delay, didn't realize this was on my plate until right now.

Looks good, with only two small nits.

Thanks so much for the contribution!

<details><summary><i>Previously, iainmcgin (Iain McGinniss) wrote…</i></summary>
> +@williamdenniss and +@stevenewright for review.
</details>


---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/OIDAuthState.m, line 467 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KJIC2DSi7KdzE4nt8rk:-KJIC2DSi7KdzE4nt8rl:909127870) ([raw file](https://github.com/openid/appauth-ios/blob/d8d673d8f714b60e55e65dd9145c9ad85e8209f9/Source/OIDAuthState.m#L467)):*
> ```objc
>     @brief Determines whether a token refresh request must be made to refresh the tokens.
>  */
> - (BOOL) isTokenValid {
> ```

Nit; please remove space between (BOOL) and "isTokenValid".

https://google.github.io/styleguide/objcguide.xml#Method_Declarations_and_Definitions

---

*[Source/OIDAuthState.m, line 476 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KJICNBzM_04nYXOan-C:-KJICNBzM_04nYXOan-D:1393963179) ([raw file](https://github.com/openid/appauth-ios/blob/d8d673d8f714b60e55e65dd9145c9ad85e8209f9/Source/OIDAuthState.m#L476)):*
> ```objc
>   }
> 
>   if (!self.accessTokenExpirationDate) {
> ```

We prefer to access backing ivars directly within the implementation, so please:

if (!_accessTokenExpirationDate) {
  return !!_accessToken;
}

return [_accessTokenExpirationDate timeIntervalSinceNow] > kExpiryTimeTolerance;

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23#-:-KJICglHiOXXKaqFmUVW:-1428301914)*
<!-- Sent from Reviewable.io -->

rebeckag commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[Source/OIDAuthState.m, line 467 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KJIC2DSi7KdzE4nt8rk:-KJeLSSS9R__6U1GXc_p:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/d8d673d8f714b60e55e65dd9145c9ad85e8209f9/Source/OIDAuthState.m#L467)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> Nit; please remove space between (BOOL) and "isTokenValid".
> 
> https://google.github.io/styleguide/objcguide.xml#Method_Declarations_and_Definitions
> 
</details>
Done.

---

*[Source/OIDAuthState.m, line 476 \[r1\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KJICNBzM_04nYXOan-C:-KJeQXFMHTywtvLLeymS:142807290) ([raw file](https://github.com/openid/appauth-ios/blob/d8d673d8f714b60e55e65dd9145c9ad85e8209f9/Source/OIDAuthState.m#L476)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> We prefer to access backing ivars directly within the implementation, so please:
> 
> if (!_accessTokenExpirationDate) {
>   return !!_accessToken;
> }
> 
> return [_accessTokenExpirationDate timeIntervalSinceNow] > kExpiryTimeTolerance;
</details>
`accessTokenExpirationDate` is a private property, with a getter method, so not just a plain ivar that can be accessed with `_accessTokenExpirationDate`. I followed the pattern for `accessToken` and `idToken` in the same file, is that ok?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

rebeckag commented:
Any update on this?

StevenEWright commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23#-:-KM64kRe17MT1_EBkTVm:1417088721)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[Source/OIDAuthState.m, line 402 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM63RABr0L3nT_sSZ1b:-KM63RABr0L3nT_sSZ1c:-263712599) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L402)):*
> ```objc
> 
> - (void)withFreshTokensPerformAction:(OIDAuthStateAction)action {
>   if (![self isTokenValid]) {
> ```

This is confusing, either the method `isTokenValid` should be named `isTokenInvalid`, or there shouldn't be a NOT operator here.  I prefer `isTokenValid` (or more accurately,`isTokenFresh`).

---

*[Source/OIDAuthState.m, line 412 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM60W_mcFA81otXkw3z:-KM60W_mcFA81otXkw4-:2048911196) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L412)):*
> ```objc
>   if (!_refreshToken) {
>     // no refresh token available and token have expired
>     [OIDErrorUtilities raiseException:kRefreshTokenRequestException];
> ```

If we want to support refresh-token-less usage, then an exception is probably wrong here.

I'd suggest adding a new error code, and calling `action(nil, nil, error)`

---

*[Source/OIDAuthState.m, line 467 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM64ABjQIqVsuChGhE_:-KM64ABjQIqVsuChGhEa:-11791601) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L467)):*
> ```objc
>     @brief Determines whether a token refresh request must be made to refresh the tokens.
>  */
> - (BOOL)isTokenValid {
> ```

should we rename to `isTokenFresh` instead?  I feel like that's what we are testing for.

---

*[Source/OIDAuthState.m, line 468 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM64IL6GhkGRpNGi8s4:-KM64IL6GhkGRpNGi8s5:250240228) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L468)):*
> ```objc
>  */
> - (BOOL)isTokenValid {
>   if (!_refreshToken) {
> ```

I'm not convinced the presence of the refresh token or lack thereof alters the validity of the current tokens.

---

*[Source/OIDAuthState.m, line 473 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM6656PAAR7FpC6vibO:-KM6656PAAR7FpC6vibP:436025794) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L473)):*
> ```objc
> 
>   if (_needsTokenRefresh) {
>     return YES;
> ```

needing a token refresh should imply the token is "invalid" not valid, so I'm expecting a NO return here.

---

*[Source/OIDAuthState.m, line 476 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM68jb8DZUpby0jKp9U:-KM68jb8DZUpby0jKp9V:1602473667) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L476)):*
> ```objc
>   }
> 
>   if (!self.accessTokenExpirationDate) {
> ```

OAuth 2.0 defines the optional field as:

```
   expires_in
         RECOMMENDED.  The lifetime in seconds of the access token.  For
         example, the value "3600" denotes that the access token will
         expire in one hour from the time the response was generated.
         If omitted, the authorization server SHOULD provide the
         expiration time via other means or document the default value
```

I think based on that, we should treat any token without an expiry time as infinitely valid (i.e. the client has to set `needsTokenRefresh` manually since the value is provided out of band).

Is that what you are doing here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

StevenEWright commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/OIDAuthState.m, line 402 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM63RABr0L3nT_sSZ1b:-KM6AC5ND4sJY0gUBPHG:224391306) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L402)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> This is confusing, either the method `isTokenValid` should be named `isTokenInvalid`, or there shouldn't be a NOT operator here.  I prefer `isTokenValid` (or more accurately,`isTokenFresh`).
</details>
If tests didn't catch this mistake, then we need to improve the tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

rebeckag commented:
Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/OIDAuthState.m, line 402 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM63RABr0L3nT_sSZ1b:-KM8RpBNoEp2E7PpjnQk:-1767757823) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L402)):*
<details><summary><i>Previously, StevenEWright (Steven E Wright) wrote…</i></summary>
> If tests didn't catch this mistake, then we need to improve the tests.
</details>
Renamed and added tests. Done.

---

*[Source/OIDAuthState.m, line 412 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM60W_mcFA81otXkw3z:-KM8SxFkOt2QeB64ztR_:-1266823048) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L412)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> If we want to support refresh-token-less usage, then an exception is probably wrong here.
> 
> I'd suggest adding a new error code, and calling `action(nil, nil, error)`
> 
> 
</details>
There is no OAuth error for this, right? I've added `OIDErrorCodeTokenRefreshError` under the `OIDGeneralErrorDomain`.

---

*[Source/OIDAuthState.m, line 467 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM64ABjQIqVsuChGhE_:-KM8U-dJN1bFiHh1fOjz:-624829205) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L467)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> should we rename to `isTokenFresh` instead?  I feel like that's what we are testing for.
</details>
Definitely! Done.

---

*[Source/OIDAuthState.m, line 468 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM64IL6GhkGRpNGi8s4:-KM8U7apoWlyTW0uvj7N:798451123) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L468)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> I'm not convinced the presence of the refresh token or lack thereof alters the validity of the current tokens.
</details>
I agree that it doesn't. The previous implementation was more in the lines of "needs token refresh?" and assuming "no" since that is impossible if we have no refresh token. But I've removed that clause. Done.

---

*[Source/OIDAuthState.m, line 473 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM6656PAAR7FpC6vibO:-KM8U6nzvZa_BcxD2x9H:-499146009) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L473)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> needing a token refresh should imply the token is "invalid" not valid, so I'm expecting a NO return here.
</details>
Done.

---

*[Source/OIDAuthState.m, line 476 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM68jb8DZUpby0jKp9U:-KM8UWZ2xlH8uhVzUxCg:-1162482214) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L476)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> OAuth 2.0 defines the optional field as:
> 
> ```
>    expires_in
>          RECOMMENDED.  The lifetime in seconds of the access token.  For
>          example, the value "3600" denotes that the access token will
>          expire in one hour from the time the response was generated.
>          If omitted, the authorization server SHOULD provide the
>          expiration time via other means or document the default value
> ```
> 
> I think based on that, we should treat any token without an expiry time as infinitely valid (i.e. the client has to set `needsTokenRefresh` manually since the value is provided out of band).
> 
> Is that what you are doing here?
</details>
Exactly. If the expiration time is not set, assume the token is still valid if it is set.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Source/OIDAuthState.m, line 412 \[r2\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KM60W_mcFA81otXkw3z:-KMcsTjuRm35dRLf_oo5:105594954) ([raw file](https://github.com/openid/appauth-ios/blob/2d3177d491f78b88682262a5a1ff53be1bd9aee9/Source/OIDAuthState.m#L412)):*
<details><summary><i>Previously, rebeckag (Rebecka Gulliksson) wrote…</i></summary>
> There is no OAuth error for this, right? I've added `OIDErrorCodeTokenRefreshError` under the `OIDGeneralErrorDomain`.
</details>
There wasn't.  Looks good.

---

*[Source/OIDAuthState.m, line 402 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KMcstZwL2unU0m2AZpq:-KMcstZwL2unU0m2AZpr:716894745) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L402)):*
> ```objc
> 
> - (void)withFreshTokensPerformAction:(OIDAuthStateAction)action {
>   if ([self isTokenFresh]) {
> ```

You replaced ` if ([self.accessTokenExpirationDate timeIntervalSinceNow] > kExpiryTimeTolerance
      && !_needsTokenRefresh) {`

With `[self isTokenFresh]`

Yet `[self isTokenFresh]` does not include the expiration date check.  That method needs to be revised.

---

*[Source/OIDAuthState.m, line 422 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2laoazEtVpfdI597t:-KN2laoazEtVpfdI597u:-1535173560) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L422)):*
> if necessary, 

"Access token is expired, first refresh the token, then perform action"

---

*[Source/OIDAuthState.m, line 483 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2m0szVdiNApe159IP:-KN2m0szVdiNApe159IQ:1260616157) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L483)):*
> !!self.accessToken

should this just be "return YES"

I don't think it's spec compliant not to have any token at all (what would be the point of the authorization request?)

---

*[Source/OIDAuthState.m, line 487 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2mBvwmorluMrnS1Sz:-KN2mBvwmorluMrnS1T-:395757205) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L487)):*
> v
replace "v" with "tokenExpired"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23#-:-KN2pm7UMNdRuylabA-_:1417088721)*
<!-- Sent from Reviewable.io -->

WilliamDenniss commented:
Looks like this is ready to go, but has some merge conflicts. Can you resolve the merge conflicts so I can get it in?

rebeckag commented:
Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/OIDAuthState.m, line 402 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KMcstZwL2unU0m2AZpq:-KPY7CJlhUtp1vvwXdzk:b-l5c8mf) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L402)):*
> You replaced if ([self.accessTokenExpirationDate timeIntervalSinceNow] > kExpiryTimeTolerance

`isTokenFresh` contains the same expiration date check.

---

*[Source/OIDAuthState.m, line 422 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2laoazEtVpfdI597t:-KPY7bhzZLMxsqHReQcC:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L422)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> > if necessary, 
> 
> "Access token is expired, first refresh the token, then perform action"
</details>
Done.

---

*[Source/OIDAuthState.m, line 483 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2m0szVdiNApe159IP:-KPY7hzO4dzGE5fc4O_b:bow5ufp) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L483)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> > !!self.accessToken
> 
> should this just be "return YES"
> 
> I don't think it's spec compliant not to have any token at all (what would be the point of the authorization request?)
</details>
With response_type='id_token', there will only be an ID Token. However, I don't know if that is necessary to support for native apps, but this at least includes that case too.

---

*[Source/OIDAuthState.m, line 487 \[r3\]](https://reviewable.io:443/reviews/openid/appauth-ios/23#-KN2mBvwmorluMrnS1Sz:-KPY7zfE3k3HoumDgiKa:b-896fix) ([raw file](https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L487)):*
<details><summary><i>Previously, WilliamDenniss (William Denniss) wrote…</i></summary>
> > v
> replace "v" with "tokenExpired"
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/openid/appauth-ios/23)*
<!-- Sent from Reviewable.io -->

ve7jtb commented:
Returning just an id_token is possible with the “id_token” response type.  That however is returned in the front channel fragment encoded.

For the code response type you always get a access token, the refresh is optional.   Now there may be no grants associated with the AT, but you still get it as it is required by OAuth. 

I guess you could define a new grant type that only returned the id_token, but that doesn’t exist currently.

So for the authorization_code grant type not getting a AT back is a AS error.

John B.
> On Aug 19, 2016, at 11:13 AM, Rebecka Gulliksson <notifications@github.com> wrote:
> 
> Review status: 0 of 3 files reviewed at latest revision, 5 unresolved discussions.
> 
> Source/OIDAuthState.m, line 402 [r3] <https://reviewable.io/reviews/openid/appauth-ios/23#-KMcstZwL2unU0m2AZpq:-KPY7CJlhUtp1vvwXdzk:b-l5c8mf> (raw file <https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L402>):
> 
> You replaced if ([self.accessTokenExpirationDate timeIntervalSinceNow] > kExpiryTimeTolerance
> 
> isTokenFresh contains the same expiration date check.
> 
> Source/OIDAuthState.m, line 422 [r3] <https://reviewable.io/reviews/openid/appauth-ios/23#-KN2laoazEtVpfdI597t:-KPY7bhzZLMxsqHReQcC:b-896fix> (raw file <https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L422>):
> Previously, WilliamDenniss (William Denniss) wrote…
> 
> Done.
> 
> Source/OIDAuthState.m, line 483 [r3] <https://reviewable.io/reviews/openid/appauth-ios/23#-KN2m0szVdiNApe159IP:-KPY7hzO4dzGE5fc4O_b:bow5ufp> (raw file <https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L483>):
> Previously, WilliamDenniss (William Denniss) wrote…
> 
> With response_type='id_token', there will only be an ID Token. However, I don't know if that is necessary to support for native apps, but this at least includes that case too.
> 
> Source/OIDAuthState.m, line 487 [r3] <https://reviewable.io/reviews/openid/appauth-ios/23#-KN2mBvwmorluMrnS1Sz:-KPY7zfE3k3HoumDgiKa:b-896fix> (raw file <https://github.com/openid/appauth-ios/blob/130bc23de35378f975b3233d7c7968eb99c5753d/Source/OIDAuthState.m#L487>):
> Previously, WilliamDenniss (William Denniss) wrote…
> 
> Done.
> 
> Comments from Reviewable <https://reviewable.io/reviews/openid/appauth-ios/23>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/openid/AppAuth-iOS/pull/23#issuecomment-241029561>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AAD4J890SuKPjLnJ7dHaVRMObC-r9keHks5qhboZgaJpZM4IkgiL>.
>

rebeckag commented on Source/OIDError.h:
```diff
@@ -118,6 +118,12 @@ typedef NS_ENUM(NSInteger, OIDErrorCode) {
+
+  /*! @brief Indicates a problem when trying to refresh the tokens.
+   */
+  OIDErrorCodeTokenRefreshError = -11,
```
Conflicts with [change](https://github.com/openid/AppAuth-iOS/pull/22/files#diff-6d5d553d8f414d088c8019134942c79bR137) introduced by #22, which should take precedence?

WilliamDenniss replied:
This PR, I'll merge it first.

rebeckag commented:
Thanks for the review and fixing the issues :)!
